platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetScoreCalculator.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (C) 2016-2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.event.tipset;
+
+import static com.swirlds.platform.Utilities.isSuperMajority;
+import static com.swirlds.platform.event.tipset.TipsetAdvancementWeight.ZERO_ADVANCEMENT_WEIGHT;
+
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.system.NodeId;
+import com.swirlds.common.system.address.AddressBook;
+import com.swirlds.platform.event.EventDescriptor;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Calculates tipset scores for events created by a node.
+ */
+public class TipsetScoreCalculator {
+
+    /**
+     * The node ID that is being tracked by this window.
+     */
+    private final NodeId selfId;
+
+    /**
+     * Builds tipsets for each event. Is maintained outside this object.
+     */
+    private final TipsetTracker tipsetTracker;
+
+    /**
+     * Tracks non-ancient events without children.
+     */
+    private final ChildlessEventTracker childlessEventTracker;
+
+    /**
+     * The current tipset snapshot. This is updated to the latest self event's tipset whenever the targeted weighted
+     * advancement between the current snapshot and the new event's tipset exceeds the threshold of 2/3 consensus weight
+     * minus the self weight.
+     */
+    private Tipset snapshot;
+
+    /**
+     * The N most recent snapshots.
+     */
+    private final Deque<Tipset> snapshotHistory = new LinkedList<>();
+
+    /**
+     * The number of snapshots to keep in {@link #snapshotHistory}.
+     */
+    private final int maxSnapshotHistorySize;
+
+    /**
+     * The total weight of all nodes.
+     */
+    private final long totalWeight;
+
+    /**
+     * The weight of the node tracked by this window.
+     */
+    private final long selfWeight;
+
+    /**
+     * The maximum possible advancement weight for an event.
+     */
+    private final long maximumPossibleAdvancementWeight;
+
+    /**
+     * The previous tipset advancement weight.
+     */
+    private TipsetAdvancementWeight previousScore = ZERO_ADVANCEMENT_WEIGHT;
```

Comment:
mixing of terms again here - weight vs. score

Reply:
Converted to the term "advancement weight" instead of "score"